### ISSUE TITLE
Consider linking against libsystem_kernel.dylib or libSystem.dylib on macOS?

### ISSUE THREAD

**Issue by copumpkin** [golang/go#17490]
As I understand it, Go currently has its own syscall wrappers for Darwin. This is explicitly against what [Apple recommends](https://developer.apple.com/library/content/qa/qa1118/_index.html), precisely because they're not willing to commit to a particular syscall ABI. This leads to issues like https://github.com/golang/go/issues/16570, and although we've been lucky in that things have generally been backward-compatible so far, there's no guarantee that it'll continue to happen. It doesn't seem inconceivable to me that we'd at some point end up having to specify "build for macOS 10.13+" vs. "build for 10.12 and below", for example.

Linking against libsystem_kernel.dylib (or the broader libSystem.dylib) would put Go back in line with Apple's recommendations for the platform, and the library is guaranteed to exist on all macOS boxes.

Apologies if this has been suggested before and people have already gone over good reasons not to do it, but I'm currently struggling to figure out what appears to be another Sierra incompatibility related to #16570 above (still haven't figured it out enough to post a reproducible bug) and am wishing we didn't have to deal with these issues.

**Comment by ianlancetaylor** [comment]
I agree that we have to do this.  I'm not sure who is actually going to do it, though.

**Comment by quentinmit** [comment]
Note that the trouble with #16570 comes from fetching the current time, and that is something that is so performance critical that we might want to continue to use our own implementation even if we use libSystem for other syscall.

**Comment by crawshaw** [comment]
We got into a discussion about how to do this over on #17200. I believe this can be done without forcing external linking. But I also don't know who is going to do it, my plate is full for the foreseeable future.

**Comment by rsc** [comment]
Agree we should do this at some point. It's not terribly high priority though.

**Comment by rsc** [comment]
This is about using the system calls. We already link against libSystem.dylib when cgo is in use.

**Comment by hirochachacha** [comment]
Hi, I made a CL for x86. https://go-review.googlesource.com/c/50290/
Could someone review it (after development cycle open) ?
If the overall design is OK, I'd like do something in go 1.10. Thank you.

**Comment by rsc** [comment]
Despite the release-blocker label I don't see this happening for Go 1.10. Moving to Go 1.11. If we really want to do this we need to make sure work starts at the beginning of the cycle.

**Comment by ianlancetaylor** [comment]
Nick Kledzik suggested in e-mail:

> There is a middle ground that some JIT creators use, which is to wrap their JITed code in a “boilerplate” mach-o.
>
> For the Go case, since you already generate the wrapper mach-o structure for executables, you expand that a bit, and standard content for linking to libSystem.dylib.  This would include the stubs (PLT) section and lazy pointer section (GOT), and the LINKEDIT stuff to bind them for all the POSIX calls Go supports.  The Go compiler then calls the stub (PLT) instead of directly inlining the syscall.
> 
> The new mach-o content would be the same for all Go executables, so you don’t need a linker. You write a C program that calls all POSIX functions Go supports and compile/link that once, and copy the stubs/lazy pointer sections and LINKEDIT content for use as your boilerplate.
> 
> This won’t solve the thread local issue, but will make Go executables true dynamic executables and free Apple to evolve the private libSystem/kernel syscall interface.

**Comment by andlabs** [comment]
Probably the wrong place to ask this, but I am curious: what does `MOVL	$0xf1, 0xf1` do to trigger a crash? Does it just get assembled to a bunch of `icebp` instructions?

**Comment by ianlancetaylor** [comment]
`MOVL $0xf1, 0xf1` assembles into exactly what it looks like: a store of a value to the memory address `0xf1`.  Since that memory address is not mapped, it causes a bus error, crashing the program.  The advantage of using `0xf1` is that the address will be printed in the stack backtrace, giving a clue where the problem happened.

**Comment by ianlancetaylor** [comment]
@randall77 Any reason to keep this issue open?

**Comment by randall77** [comment]
Yes, I still need to do x/sys/unix and vendor it into the stdlib.
I think everything else is done.
